cmd/virt-handler/virt-handler.go

```diff
@@ -68,19 +71,41 @@ func main() {
 		panic(err)
 	}
 
+	// Create event recorder
+	coreClient, err := kubecli.Get()
+	if err != nil {
+		panic(err)
+	}
+	broadcaster := record.NewBroadcaster()
+	broadcaster.StartRecordingToSink(&kubecorev1.EventSinkImpl{Interface: coreClient.Events(api.NamespaceDefault)})
+	recorder := broadcaster.NewRecorder(kubev1.EventSource{Component: "virt-handler", Host: *host})
+
 	domainListWatcher := virtcache.NewListWatchFromClient(domainLWConn, libvirtapi.VIR_DOMAIN_EVENT_ID_LIFECYCLE)
 
 	_, domainController := kubecli.NewInformer(domainListWatcher, &libvirt.Domain{}, 0, kubecli.ResourceEventHandlerFuncs{
 		AddFunc: func(obj interface{}) error {
-			fmt.Printf("Domain ADDED: %s: %s\n", obj.(*libvirt.Domain).GetObjectMeta().GetName(), obj.(*libvirt.Domain).Status.Status)
+			domain := obj.(*libvirt.Domain)
+			// TODO this just demonstrates how to do event logging. More fine grained differentiations are necessary here.
+			// TODO For now we use an empty reference for event logging, normally whenever an event is logged, an update on the VM object itself is also part of this
+			if domain.Status.Status == libvirt.Running {
+				recorder.Event(NewVMReferenceFromDomain(domain), kubev1.EventTypeNormal, "Started", "VM started")
```

Comment:
> When looking into the kubelet code it looks like it does the same.

Ok, checked it again, it is not consistent. Some pod events are "at-most-once", container events are normally "at-most-once" except when the kubelet is restarted. I think it is fine if we send the same events multiple times. There are protection mechanisms in place to avoid overwhelming the api server.